// ng-sticky/ng-sticky.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/common'], factory) :
	(factory((global['ng-sticky'] = global['ng-sticky'] || {}),global._angular_core,global._angular_common));
}(this, (function (exports,_angular_core,_angular_common) { 'use strict';

var NgStickyDirective = (function () {
    /**
     * @param {?} el
     * @param {?} render
     */
    function NgStickyDirective(el, render) {
        this.el = el;
        this.render = render;
        this.sticked = true;
        this.selectedOffset = 0;
        this.windowOffsetTop = 0;
        this.addClass = 'fixed';
        this.offSet = 0;
        this.selectedOffset = this.el.nativeElement.offsetTop;
    }
    /**
     * @return {?}
     */
    NgStickyDirective.prototype.addSticky = function () {
        this.sticked = true;
        this.el.nativeElement.style.position = 'fixed';
        this.el.nativeElement.style.top = this.offSet + 'px';
        this.render.addClass(this.el.nativeElement, this.addClass);
    };
    /**
     * @return {?}
     */
    NgStickyDirective.prototype.removeSticky = function () {
        this.sticked = false;
        this.el.nativeElement.style.position = '';
        this.render.removeClass(this.el.nativeElement, this.addClass);
    };
    /**
     * @return {?}
     */
    NgStickyDirective.prototype.onWindowScroll = function () {
        var /** @type {?} */ offset = this.el.nativeElement.offsetTop;
        this.windowOffsetTop = window.pageYOffset || document.documentElement.scrollTop || document.body.scrollTop || 0;
        if (this.selectedOffset === 0) {
            this.selectedOffset = offset;
        }
        if (this.sticked === false) {
            this.selectedOffset = offset;
        }
        if ((this.windowOffsetTop + this.offSet) > this.selectedOffset) {
            this.addSticky();
        }
        else {
            this.removeSticky();
        }
    };
    return NgStickyDirective;
}());
NgStickyDirective.decorators = [
    { type: _angular_core.Directive, args: [{
                selector: '[ng-sticky]'
            },] },
];
/**
 * @nocollapse
 */
NgStickyDirective.ctorParameters = function () { return [
    { type: _angular_core.ElementRef, },
    { type: _angular_core.Renderer2, },
]; };
NgStickyDirective.propDecorators = {
    'addClass': [{ type: _angular_core.Input },],
    'offSet': [{ type: _angular_core.Input },],
    'onWindowScroll': [{ type: _angular_core.HostListener, args: ["window:scroll", [],] },],
};

var NgStickyModule = (function () {
    function NgStickyModule() {
    }
    /**
     * @return {?}
     */
    NgStickyModule.forRoot = function () {
        return {
            ngModule: NgStickyModule
        };
    };
    return NgStickyModule;
}());
NgStickyModule.decorators = [
    { type: _angular_core.NgModule, args: [{
                imports: [
                    _angular_common.CommonModule
                ],
                declarations: [
                    NgStickyDirective
                ],
                exports: [
                    NgStickyDirective
                ]
            },] },
];
/**
 * @nocollapse
 */
NgStickyModule.ctorParameters = function () { return []; };

exports.NgStickyModule = NgStickyModule;
exports.NgStickyDirective = NgStickyDirective;

Object.defineProperty(exports, '__esModule', { value: true });

})));
